### PR TITLE
add userdata template

### PR DESCRIPTION
**Pull request by LittleBoxOfSunshine**
# PR Checklist

Check these items before submitting a PR... 

[Contribution Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md)

[Best Practice Guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md)

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

## Changelog

Adds an example template for the user data feature.

### CODE REVIEW

File: 101-vm-userdata/README.md [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+![Best Practice Check](https://azurequickstartsservice.blob.core.windows.net/badges/101-vm-userdata/BestPracticeResult.svg)
+![Cred Scan Check](https://azurequickstartsservice.blob.core.windows.net/badges/101-vm-userdata/CredScanResult.svg)
+
+[![Deploy To Azure](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.svg?sanitize=true)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-vm-userdata%2Fazuredeploy.json)  [![Visualize](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/visualizebutton.svg?sanitize=true)](http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-vm-userdata%2Fazuredeploy.json)
+
+This template allows you to create a Virtual Machine with User Data. This template also deploys a Storage Account, Virtual Network, Public IP addresses and a Network Interface.
+
+You can verify the Custom Data in the file `/var/lib/cloud/instance/user-data.txt`.
```
Comment (roipmsft) [inline 101-vm-userdata/README.md:16]: this template is not related to customData, only userData, right?
so this line should be dropped?

File: 101-vm-userdata/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+        "hardwareProfile": {
+          "vmSize": "[parameters('vmSize')]"
+        },
+        "osProfile": {
+          "computerName": "[variables('vmName')]",
+          "adminUsername": "[parameters('adminUsername')]",
+          "adminPassword": "[parameters('adminPasswordOrKey')]",
+          "linuxConfiguration": "[if(equals(parameters('authenticationType'), 'password'), json('null'), variables('linuxConfiguration'))]"
+        },
+		"userData": "[base64(parameters('userData'))]",
```
Comment (roipmsft) [inline 101-vm-userdata/azuredeploy.json:216]: nit: indented too much

File: 101-vm-userdata/README.md (hunk 1 above)
Comment (LittleBoxOfSunshine) [reply 101-vm-userdata/README.md:16]: Missed that one. There's no equivalent file for user data correct?

Comment (roipmsft) [reply 101-vm-userdata/README.md]: For right now (April 2021), at least, UserData will not be copied to a local file, so this template should not mention any local file.
For this GA, UserData is not being passed to CloudInit. This GA only enables UserData showing up in IMDS.

However, In July 2021 (or later), CloudInit is supposed to start using UserData.
Maybe CloudInit will copy UserData  to `/var/lib/cloud/instance/user-data.txt` in the future?

I am reading https://cloudinit.readthedocs.io/en/latest/topics/faq.html, but not clear on the details/expectations of CloudInit.

File: 101-vm-userdata/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+      "type": "string",
+      "metadata": {
+        "description": "User name for the Virtual Machine."
+      }
+    },
+    "userData": {
+      "type": "string",
+      "defaultValue": "echo userData",
+      "metadata": {
+        "description": "String passed down to the Virtual Machine."
```
Comment (roipmsft) [inline 101-vm-userdata/azuredeploy.json:21]: nit: description is kind of vague. maybe Ning has better description than just 'string'

Comment (LittleBoxOfSunshine) [reply 101-vm-userdata/azuredeploy.json:21]: Fair point. Talked with Ning and she suggested keeping it as is.

File: 101-vm-userdata/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+          }
+        },
+        "networkProfile": {
+          "networkInterfaces": [
+            {
+              "id": "[resourceId('Microsoft.Network/networkInterfaces',variables('nicName'))]"
+            }
+          ]
+        },
+        "diagnosticsProfile": {
```
Comment (bmoore-msft) [inline 101-vm-userdata/azuredeploy.json:221]: remove this property (it's the default)

File: 101-vm-userdata/azuredeploy.parameters.json [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
```
Comment (bmoore-msft) [inline 101-vm-userdata/azuredeploy.parameters.json:4]: remove the parameters that have defaultValues

File: 101-vm-userdata/azuredeploy.parameters.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+      "value": "Standard_D2_v2"
+    },
+    "ubuntuOSVersion": {
+      "value": "18.04-LTS"
+    },
+    "adminPasswordOrKey": {
+      "value": "GEN-SSH-PUB-KEY"
+    },
+    "location": {
+      "value": "northcentralus"
```
Comment (bmoore-msft) [inline 101-vm-userdata/azuredeploy.parameters.json:24]: if you need to target a specific location you'll need to add a separate param file for USGov: https://github.com/Azure/azure-quickstart-templates/tree/master/1-CONTRIBUTION-GUIDE#cloud-specific-parameter-files

File: 101-vm-userdata/metadata.json [hunk 7]
```diff
@@ -0,0 +1,9 @@
+{
+  "$schema": "https://aka.ms/azure-quickstart-templates-metadata-schema#",
+  "type": "QuickStart",
+  "itemDisplayName": "Deploy a Virtual Machine with User Data",
+  "description": "This template allows you to create a Virtual Machine with User Data passed down to the VM. This template also deploys a Storage Account, Virtual Network, Public IP addresses and a Network Interface.",
```
Comment (bmoore-msft) [inline 101-vm-userdata/metadata.json:5]: remove storageAccount

File: 101-vm-userdata/README.md [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+[![Deploy To Azure](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.svg?sanitize=true)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-vm-userdata%2Fazuredeploy.json)  [![Deploy To AzureGov](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazuregov.svg?sanitize=true)](https://portal.azure.us/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-vm-userdata%2Fazuredeploy.json)  [![Visualize](https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/visualizebutton.svg?sanitize=true)](http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F101-vm-userdata%2Fazuredeploy.json)
+
+This template allows you to create a Virtual Machine with User Data. This template also deploys a Storage Account, Virtual Network, Public IP addresses and a Network Interface.
```
Comment (bmoore-msft) [inline 101-vm-userdata/README.md:14]: remove storageAccount

**Review by bmoore-msft — CHANGES_REQUESTED** [review]

File: 101-vm-userdata/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,244 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "dnsLabelPrefix": {
+      "type": "string",
+      "metadata": {
```
[the remark this answers is not in the document]
Comment (LittleBoxOfSunshine) [reply 101-vm-userdata/azuredeploy.json:7]: Should I also be removing the reference to "Storage Account" in the description here?

[the remark this answers is not in the document]
Comment (LittleBoxOfSunshine) [reply 101-vm-userdata/azuredeploy.json:7]: Removed reference, this is only used for the IP address now.

File: 101-vm-userdata/azuredeploy.parameters.json (hunk 6 above)
Comment (LittleBoxOfSunshine) [reply 101-vm-userdata/azuredeploy.parameters.json:18]: This PR will need to wait until at least one gov region has fully deployed in that case, at which point it may just be worth leaving this as a draft until everything is fully deployed. The short term goal here is just to have some example to link to in announcement documentation. We'll probably bundle the template with our code examples as a short term solution until this is ready.
